src/go/log/zap/zap.go

```diff
@@ -0,0 +1,156 @@
+// Copyright 2021 The Magma Authors.
+//
+// This source code is licensed under the BSD-style license found in the
+// LICENSE file in the root directory of this source tree.
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package zap
+
+import (
+	"github.com/pkg/errors"
+	"go.uber.org/zap"
+
+	"github.com/magma/magma/log"
+)
+
+type printer struct {
+	print  func(args ...interface{})
+	printf func(format string, args ...interface{})
+}
+
+func (p *printer) Print(args ...interface{}) {
+	p.print(args...)
+}
+
+func (p *printer) Printf(format string, args ...interface{}) {
+	p.printf(format, args...)
+}
+
+type Logger struct {
+	names []string
+	*zap.Logger
+	zap.Config
+
+	printers map[log.Level]*printer
+}
+
+var _ log.Logger = (*Logger)(nil)
```

Comment:
I am not piecing together at a glance what is happening here. Will try to reconstruct by looking at references to log.Logger.

Reply:
Nope, not getting it =D I see no log package .Logger object? I see a Logger within the `log.ManagedLogger` type. How is it that `log.Logger` here is setting the logger pointer inside the ManagedLogger? I'm missing something.

Reply:
Secondary question is how does this presently-nil pointer get initialized to our zap logger? I see nothing in `log/zap/zap.go` doing so.

Reply:
this is a golang idiosyncratic compile-type interface assertion.

https://commaok.xyz/post/compile-time-assertions/

note that not everyone likes these because some prefer conformance to an interface to be optional by design. i.e. if it's necessary for a struct to conform to an interface, then it'll be used as such somewhere else in the code. If you "accidentally" stop conforming, but your tests pass, then there really isn't a problem.

Reply:
I need a `far out, dude` emoji =D